app/cdash/include/login_functions.php

```diff
@@ -0,0 +1,584 @@
+function databaseAuthenticate($email, $password, $SessionCachePolicy, $rememberme)
+{
+    global $loginerror, $pdo;
+    $loginerror = '';
+
+    $config = Config::getInstance();
+    $service = ServiceContainer::getInstance();
+    /** @var Session $session */
+    $session = $service->get(Session::class);
+    $user = $service->create(User::class);
+
+    $userid = $user->GetIdFromEmail($email);
+    if (!$userid) {
+        $loginerror = 'Wrong email or password.';
+        return false;
+    }
+
+    $user->Id = $userid;
+    $user->Fill();
+
+    // Check if the account is locked out.
+    if (accountIsLocked($userid)) {
+        return false;
+    }
+
+    if ($password === null && $config->get('CDASH_EXTERNAL_AUTH')) {
+        // External authentication succeeded.
+        // Create the session array.
+        $session->setSessionVar('cdash', [
+            'login' => $email,
+            'passwd' => null,
+            'ID' => $session->getSessionId(),
+            'valid' => 1,
+            'loginid' => $userid,
+        ]);
+
+        return true;
+    } else {
+        $success = false;
+        if (password_verify($password, $user->Password)) {
+            $success = true;
+        } elseif (md5($password) == $user->Password) {
+            // Re-hash this password using an algorithm that's more secure than md5.
+            // Do not attempt this before the database has been upgraded
+            // to accommodate the increased length of this field.
+            $config = Config::getInstance();
+            $db_check = true;
+            if ($config->get('CDASH_DB_TYPE') != 'pgsql') {
+                $table_name = qid('user');
+                $select = $pdo->query("SELECT password FROM $table_name LIMIT 1");
+                $meta = $select->getColumnMeta(0);
+                if ($meta['len'] < 255) {
+                    $db_check = false;
+                }
+            }
+            if ($db_check) {
+                $passwordHash = User::PasswordHash($password);
+                if ($passwordHash === false) {
+                    $loginerror = 'Failed to hash password.  Contact an admin.';
+                } else {
+                    $user->Password =  $passwordHash;
+                    $user->Save();
+                }
+            }
+            $success = true;
+        }
+
+        if ($success) {
+            // Authentication is successful.
+            if ($rememberme) {
+                $key = generate_password(32);
+                $session->setRememberMeCookie($user, $key);
+            }
+
+            $session->start($SessionCachePolicy);
+            $session->setSessionVar('cdash', [
+                'login' => $email,
+                'passwd' => $user->Password,
+                'ID' => $session->getSessionId(),
+                'valid' => 1,
+                'loginid' => $userid,
+            ]);
+
+            // TODO: responsibility probably belongs to user model
+            checkForExpiredPassword();
```

Comment:
Can we call `$user->hasExpiredPassword()` here and remove `checkForExpiredPassword()` from this file?